feat(markdown): configured aliases for fenced code block languages

bennypowers opened the pull request:
Closes #2131  

This PR, in the same vein as #4522, hard codes configuration for certain language aliases, namely `sh => bash`, `js => javascript` and `ts => typescript`

before:

![Screenshot from 2023-04-15 20-51-06](https://user-images.githubusercontent.com/1466420/232245403-e5136bb8-353e-476b-b4ed-d52376d03bc4.png)

after:
![Screenshot from 2023-04-15 20-51-12](https://user-images.githubusercontent.com/1466420/232245408-fdc27df6-dad1-48ad-bf62-5c8f335bc846.png)

amaanq commented:
This is something I've wanted before. The solution I had in mind was to use a grammars package.json, as the injection regex field contains this info on how to match it in markdown or any other similar style. Instead, could we rename the table to injection_language_tags? and I'll add as many as I can find from my grammars.

bennypowers commented:
Yeah that sounds good I'm not married to the local config in the file, but this did work for us in the past (#4522)

Perhaps we could merge this commit so that users are unblocked, and then go back and refactor the precise location of the config table.

In the mean time, I renamed the var as you suggested and moved the config to the top of the file.

amaanq commented:
Lmk your thoughts now, I fixed up the unnecessary commits & added missing fts

bennypowers commented on queries/markdown/injections.scm:
```diff
@@ -1,8 +1,12 @@
+  (code_fence_content)
+    @content
+    (#set-lang-from-markdown-fence-tag! @_lang)
+    ; (#exclude_children! @content))
```
should either replace or remove

bennypowers commented:
yeah @amaanq looks good habibi

theHamsta commented on lua/nvim-treesitter/query_predicates.lua:
```diff
@@ -1,5 +1,30 @@
+  sh = "bash",
+  ungram = "ungrammar",
+  uxn = "uxntal",
+}
```
shouldn't these be user configurable?

Also the interpretation of injection alias should be upstreamed at some time and maybe even resolve that abreviations by default. It's fine to have them as a experiment here. But when Neovim is trying to import our markdown queries they will need to support that directive in some way or have a more general solution

amaanq replied:
I was thinking this could be a temporary helpful bandaid-like patch until Neovim implements something better.

bennypowers replied:
I believe @amaanq 's suggestion was in the right direction, and that they should be configurable as well

So ideally, priority would be

0. Hard coded aliases
1. aliases derived from config in parser package
2. User aliases

I also think that each of those could probably be a subsequent PR

theHamsta replied:
yeah, this can also happen in a follow-up PR. I was just thinking of giving the user a chance to manipulate/extend the hardcoded table.

mmbosschaert replied:
Has the feature to allow users to manipulate or extend the hardcoded table been implemented yet?

LuisA92 commented:
+1 for user configuration. I've been using [MyST](https://myst-parser.readthedocs.io/en/latest/syntax/code_and_apis.html) flavored markdown to write scientific documents. Some code blocks begin with the syntax ` ```{code-cell} <language>`. Currently, treesitter recognizes `{code-cell}` as the language.

lucario387 commented:
Is there anyway to access vim.filetype table in 0.8 (and 0.9)? That way mappings like this aren't necessary I'd assume

amaanq commented:
> Is there anyway to access vim.filetype table in 0.8 (and 0.9)? That way mappings like this aren't necessary I'd assume

No, it's a local table, so not accessible w/ vim.filetype

ObserverOfTime commented:
This is also useful for rst.

theHamsta commented:
> Is there anyway to access vim.filetype table in 0.8 (and 0.9)? That way mappings like this aren't necessary I'd assume

Plenary also has a list of extensions https://github.com/nvim-lua/plenary.nvim/blob/master/data/plenary/filetypes/base.lua. It would be nice to hook  use/abuse as it already has a mechanism to register for users extensions. But extensions are not exactly the same as aliases (file extensions are only subset of possible aliases for a language). Maybe it should be an additional option lang.register instead? I don't have a clear opinion on that yet.

Is there already an Neovim upstream issue on this topic? It would be good to answer whether we should go with an experimental solution for some injections ahead and later upstream it or whether there could be an upstream solution from the beginning

lucario387 commented:
`vim.filetype.match` is probably the closest we can get to extension->filetype mapping. And from this we can map it out again

From what I can see, we already have somewhat of a map from  filetype->parser in https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/parsers.lua#L28. So adding more should not be much of an issue (i.e. if `injection.language` doesn't match to a parser name right away, do a `vim.filetype.match` to get ft -> getting parser name again)

bennypowers commented:
> https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/parsers.lua#L28.

Looking through that map it seems like it's not quite complete yet, and wouldn't serve here without some additional work, which I maintain is outside the scope of this PR. I will check out the `vim.filetype.match` suggestion though, which might be in scope here 

Perhaps after this merges, a subsequent PR (with regression tests!) should establish a central place for things like mimetype or alias to filetype. Perhaps this should be it's own module that exposes some `get_language_by_extention_or_mimetype`, or whatever.

theHamsta commented:
> > https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/parsers.lua#L28.
> 
> Looking through that map it seems like it's not quite complete yet, and wouldn't serve here without some additional work, which I maintain is outside the scope of this PR. I will check out the `vim.filetype.match` suggestion though, which might be in scope here
> 
> Perhaps after this merges, a subsequent PR (with regression tests!) should establish a central place for things like mimetype or alias to filetype. Perhaps this should be it's own module that exposes some `get_language_by_extention_or_mimetype`, or whatever.
> 
> edit: as you can see, vim.filetype.match gives incomplete results:
> 
> ```lua
> {
>   ex = {
>     aliased_to = "elixir"
>   },
>   fc = {
>     aliased_to = "func",
>     ft_match = "func"
>   },
>   fir = {
>     aliased_to = "firrtl",
>     ft_match = "firrtl"
>   },
>   ha = {
>     aliased_to = "hare",
>     ft_match = "hare"
>   },
>   hs = {
>     aliased_to = "haskell",
>     ft_match = "haskell"
>   },
>   js = {
>     aliased_to = "javascript",
>     ft_match = "javascript"
>   },
>   kt = {
>     aliased_to = "kotlin",
>     ft_match = "kotlin"
>   },
>   ml = {
>     aliased_to = "ocaml",
>     ft_match = "ocaml"
>   },
>   pl = {
>     aliased_to = "perl"
>   },
>   py = {
>     aliased_to = "python",
>     ft_match = "python"
>   },
>   rs = {
>     aliased_to = "rust",
>     ft_match = "rust"
>   },
>   sh = {
>     aliased_to = "bash"
>   },
> ```

When we merge this we will get a PR for every desired change in the initial list you provided.

With `vim.filetype.match(filename = "foo."..extension)` we would cover most of your list. Mimetypes if needed can be their own file if `filetype.match` doesn't give any result.

bennypowers commented:
@theHamsta is this more to your liking?

lucario387 commented on lua/nvim-treesitter/query_predicates.lua:
```diff
@@ -1,5 +1,25 @@
+  ts = "typescript",
+}
+
+local function get_language_name_from_fenced_code_block_tag(injection_alias)
```
can this be named something else, like `get_parser_from_extension` or such?

bennypowers replied:
it's not an extension, though. GFM spec calls it an ["info string"](https://github.github.com/gfm/#info-string)

bennypowers replied:
how's that?

lucario387 replied:
Agreed, but it can be changed into other names, because usage of this can be, for example, in vue/svelte `lang="..."` attributes, where having this directive will be useful. From the viewpoint of using it in otherplaces, it's easier to follow if the functions related to this directive is named to reflect its usage

lucario387 commented on queries/markdown/injections.scm:
```diff
@@ -1,8 +1,11 @@
-  (#not-match? @language "elm")
-  (code_fence_content) @content (#exclude_children! @content))
+    (language) @_lang)
+  (#not-match? @_lang "elm") ; prevent segfault when using elm parser
```
Sorry for not mentioning, but is there any other parser that causes segfault here? Aside from this, I agree with the change

amaanq replied:
Don't think so, Nim might if we add it

theHamsta commented on lua/nvim-treesitter/query_predicates.lua:
```diff
@@ -119,19 +139,17 @@ query.add_predicate("has-type?", function(match, _pattern, _bufnr, pred)
 ---@return boolean|nil
-query.add_directive("set-lang-from-mimetype!", function(match, pattern, bufnr, predicate, metadata)
-  local capture_id = predicate[2]
+query.add_directive("set-lang-from-mimetype!", function(match, _, bufnr, pred, metadata)
```
Maybe set language from alias?

theHamsta reviewed (approved):
Would be ok for me as a first draft though this needs to be configurable (not also by extensions at some time)

bennypowers commented:
agreed. Looking just now through lua/nvim-treesitter/configs.lua, it's not obvious to me where this config would live, so i think it's best to open a new issue and gain consensus before attempting to make this configurable

theHamsta commented:
Should we document this new directive somewhere? I think even the other predicates/directives of nvim-treesitter are missing documentation. Can also be left for follow-up

lucario387 commented:
There will need to have a follow up PR for other directives, and that's about it

WhyNotHugo commented:
Thanks a lot for this patch!

AckslD commented:
Hope it's okay to ask here: is there a follow-up issue/PR for user configuration that I can subscribe to?

clason commented:
`:h vim.treesitter.language.register`
